docs/developer-guide/custom-widgets/universal-widgets.md

```diff
@@ -0,0 +1,339 @@
+# Universal Widgets
+
+Widgets in deck.gl allow developers to create custom UI elements that are deeply integrated into the deck.gl rendering system. This guide covers the steps to implement widgets that are framework-agnostic, ensuring compatibility across various applications.
+
+## Implementing the Widget Lifecycle Functions
+
+The widget lifecycle functions define how a widget initializes, updates, and cleans up its integration with deck.gl.
+
+### Adding a Widget
+
+[`onAdd({deck, viewId}): HTMLElement?`](../../api-reference/core/widget.md#onadd) - This method provides deck.gl with the root DOM element of your widget. This element is positioned based on `placement` and `viewId` members.
+
+```ts
+import { type Widget } from '@deck.gl/core'
+
+class AwesomeWidget implements Widget {
+  onAdd({ deck, viewId }) {
+    const element = document.createElement('div');
+    // Initialize and style your element
+    return element
+  }
+}
+```
+
+### Updating Properties
+
+[`setProps(props)`](../../api-reference/core/widget.md#setprops) - This method is called whenever widget properties are updated. Use this to apply changes dynamically.
+
+```ts
+class AwesomeWidget implements Widget {
+  setProps(props: Partial<ZoomWidgetProps>) {
+    Object.assign(this.props, props);
+  }
+}
+```
+
+### Handling Viewport Changes
+
+[`onViewportChange(viewport: Viewport)`](../../api-reference/core/widget.md#onviewportchange) - Widgets can listen to viewport updates via the `onViewportChange` method. You can target specific viewports or listen to all. 
+
+Specific viewport:
+
+```ts
+class AwesomeWidget implements Widget {
+  viewId = 'minimap'
+
+  onViewportChange(viewport: Viewport) {
+    // Handle updates for the "minimap" viewport
+  }
+}
+```
+
+All viewports:
+
+```ts
+class AwesomeWidget implements Widget {
+  viewId = null
+
+  onViewportChange(viewport: Viewport) {
+    // Handle updates for all viewports
+  }
+}
+```
+
+### Redrawing Widget
+
+[`onRedraw({viewports, layers})`] - This method is invoked during rendering. This is useful for responding to changes in layers or viewports, such as a `HeatmapLayer`'s `colorRange` prop.
+
+```ts
+class AwesomeWidget implements Widget {
+  onRedraw({ viewports, layers }) {
+    // Update widget visuals
+  }
+}
+```
+
+### Removing Widget
+
+[`onRemove()`](../../api-reference/core/widget.md#onremove) - If implemented, this method is called when your widget is removed. This is a good time to clean up resources.
+
+```ts
+class AwesomeWidget implements Widget {
+  onRemove() {
+    // Cleanup code
+  }
+}
+```
+
+-----
+
+## Handling View Interaction Events
+
+A widget can update in response to a user interacting with the deck.gl view the widget is attached to. 
```

Comment:
This is easily the most important section and needs its own page IMHO. (I do understand that we have work to do here.)

Reply:
I've been struggling to port the graph layers controller React component to a widget, I was looking for docs on how to implement zooms and pans

https://github.com/visgl/deck.gl-community/blob/master/modules/graph-layers/src/widgets/view-control-widget.tsx#L325